### ISSUE TITLE
Custom attributes that have strings ending in 0 don't roundtrip

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtime#49756]
```csharp
[My("hello\0")]
class MyAttribute : Attribute
{
    public MyAttribute(params object[] arr)
    {
    }
}
```

Reading the above with System.Reflection.Metadata's CustomAttribute.DecodeValue claims that the argument to My is `hello` instead of `hello\0`.

I didn't debug into it, but from source code inspection this is caused by the following (Roslyn specific?) compat quirk:

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobReader.cs#L594-L596

Would be nice to be able to opt out of this.

System.Reflection.Metadata is currently unable to decode a lot of custom attributes in this very repo because putting a `\0` is popular in various InlineDatas:

https://github.com/dotnet/runtime/blob/128c757a3ac580829d08c2a60aa205559ccde8f9/src/libraries/Common/tests/Tests/System/StringTests.cs#L78

Cc @tmat

**Comment by tmat** [comment]
/cc @jaredpar

Not sure what was the legacy compiler compat reason that lead to adding this. 

I think it would make sense to remove the trimming from `BlobReader.ReadSerializedString` in .NET 6 and add it to Roslyn instead. Or perhaps even removing it from Roslyn?

It'd be a breaking change but it doesn't seem like a significant one - since `\0` can still be returned if it is in the middle of the string the caller must be able to handle strings with `\0` anyways.

**Comment by jaredpar** [comment]
@AlekseyTs do you know of any specific cases that drove this compat quirk? 

If not I think we could consider fixing for .NET 6. As @tmat mentioned if we find a real case that drove this we could consider putting the patch in the roslyn code base.

**Comment by AlekseyTs** [comment]
I do not know the specific case, but probably changing the behavior will break some unit-test with specific scenario.

**Comment by AlekseyTs** [comment]
Native compilers behavior was probably caused by the fact that C++ strings are null-terminated and everything after `\0` was ignored.

**Comment by MichalStrehovsky** [comment]
Actually - does Roslyn depend on this at all? This appears to be used in two spots: `ReadSerializedString`, the public API, and as an implementation detail of the custom attribute decoder.

I found some instances of `ReadSerializedString` in the Roslyn repo that appear to be test-only. Then I searched for `ICustomAttributeTypeProvider` that would indicate the attribute decoder is being used, and found nothing? Could it be that this has no impact on Roslyn?

If so, https://github.com/dotnet/runtime/pull/50088 should be hopefully a trivial review.
